### ISSUE TITLE
`~` and `+` should share a hanging indent for `indentation_linter()`

### ISSUE THREAD

**Issue by AshesITR** [r-lib/lintr#1718]
Follow-Up to #1411

This seems like a peculiarity in the way `~` is parsed. Good catch, but should be a follow up I think.
See
``` r
lintr::lint(text = "y ~\n  x1 +\n  x2", linter = lintr::indentation_linter())
#> <text>:3:2: style: [indentation_linter] Indentation should be 4 spaces but is 2 spaces.
#>   x2
#>  ^~~
lintr::lint(text = "y +\n  x1 +\n  x2", linter = lintr::indentation_linter())
```

<sup>Created on 2022-10-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

_Originally posted by @AshesITR in https://github.com/r-lib/lintr/pull/1411#discussion_r996412841_

**Comment by MichaelChirico** [comment]
This affects all mixed infix operators, is there anything special about `(~, +)` that we should treat them the same, or do no infixes induce a new indentation level?

https://github.com/r-lib/lintr/blob/9af53f1795f0776ae4cd790bfda1f205daec418c/R/indentation_linter.R#L192

e.g.

```r
lintr::lint(text = "y ~\n  x1 $\n  x2", linter = lintr::indentation_linter())
lintr::lint(text = "y |\n  x1 &\n  x2", linter = lintr::indentation_linter())
```

Generally:

```r
infix_tokens = environment(indentation_linter())$infix_tokens
infix_strings = infix_metadata$string_value[match(infix_tokens, infix_metadata$xml_tag)]
pairs = expand.grid(s1 = infix_strings, s2 = infix_strings, stringsAsFactors=FALSE)

linter = indentation_linter()
lints <- lapply(seq_len(nrow(pairs)), \(i) lint(sprintf("y %s\n  x1 %s\n  x2\n", pairs$s1[i], pairs$s2[i]), linter))
pairs$lint_type = pairs$lint_type = sapply(lints, \(l) if (length(l)) l[[1]]$type else "none")

table(pairs$lint_type)
# error  none style 
#   260   415   225 

lint_pairs <- subset(pairs, lint_type == "style")
lint_pairs[sample(nrow(lint_pairs), 10), c("s1", "s2")]
#     s1 s2
# 498  =  =
# 665 >=  ^
# 341  | &&
# 15  ->  +
# 754  >  :
# 670  &  ^
# 881  |  $
# 609 !=  /
# 694  >  @
# 419  ? <-
```

The errors come from operators that can't be chained and/or mixed (`c("|>", "::", ":::", "!")` and `c(">", "<", ">=", "<=", "!=", "==")`).

Note that `=` and `^` also generate lints against themselves (`y ^\n  x1 ^\n  x2`).

**Comment by AshesITR** [comment]
I'm not quite sure tbh. Given some thought, we should probably raise this with `{styler}` instead.
